open-apparel-registry/open-apparel-registry
# Add user-customizable API throttling

## Overview

Adds a memcached caching backend, and uses it to add rate throttling using the DRF built-in rate throttling, along with a custom throttle class to enable us to configure throttling at a per-user level in the Django admin.

Connects #2092

## Demo

Error response returning after throttling:
![localhost_8081_api_facilities_](https://user-images.githubusercontent.com/4432106/188977044-d7e7d238-a9f1-4948-9336-feaa125cd499.png)

Admin fields to configure throttling per user:
![localhost_8081_admin_api_user_1_change_ (2)](https://user-images.githubusercontent.com/4432106/188977047-335db444-be1b-4780-81e5-3e191d705185.png)


## Notes

This doesn't add Elasticache to our AWS environment, so this will use the default local memory caching on staging for API throttling, until that is implemented.

I wasn't entirely sure what endpoints to throttle or what default rates to use - I'd appreciate feedback here, preferably based on OAR usage statistics if we have them handy, I started with what seemed like conservative defaults and only unthrottled GET endpoints that seemed fairly simplistic - we could probably be more aggressive in what we allow.

## Testing Instructions

* `scripts/update`
* Try to trigger rate limiting
* Set your user's rate limits in the admin and verify they've taken effect

## Checklist
